drgn/helpers/linux/fs.py

```diff
@@ -120,6 +151,19 @@ def path_lookup(
                 failed_path = os.fsdecode(b"/".join(components[: i + 1]))
                 raise Exception(f"could not find {failed_path!r} in dcache")
             mnt, dentry = _follow_mount(mnt, dentry)
+            if dentry.d_inode and is_inode_symlink(dentry.d_inode):
+                s = get_link(dentry.d_inode)
+                if not s:
+                    raise Exception(f"could not get link from inode")
+
+                # full symbol link target path
+                if s[0:1] == b'/':
+                    link_target = s
+                else:
+                    solved_path = b"/" + b"/".join(components[0:i+1])
+                    link_target = solved_path + "/" + s
```

Comment:
This will redo the entire lookup up to the symlink, right? That sounds pretty inefficient (and racy on live systems), especially if there are multiple symlinks in the path traversal.

There are probably multiple ways to avoid this, but the one that comes to mind is to keep a stack of names being resolved and avoid recursion. Something like:

```python3
class PathLookupName:
    def __init__(self, name: str) -> None:
        self.components = name.split(b"/")
        self.index = 0

name_stack = [PathLookupName(os.fsencode(path))]
while name_stack:
    name = name_stack[-1]
    if name.index >= len(name.components):
        name_stack.pop()
        continue

    component = name_stack.components[name.index]
    name.index += 1

    ... handle component ...

    if is_symlink(...):
        link_target = ...
        name_stack.push(PathLookupName(link_target))
        if link_target.startswith(b"/"):
            # Restart at the root.
            mnt = root_mnt
            dentry = root_dentry
```

I _think_ that general idea will work. We also need to be careful about cycles (the kernel handles this by returning `ELOOP` if it encounters too many symlinks.)

This is going to need lots of test cases :smile: